docs: platform portability and infrastructure documentation (#91237)
vercel/next.js: docs/01-app/02-guides/cdn-caching.mdx

related:
  description: Related guides and references.
  links:
    - app/guides/deploying-to-platforms
    - app/guides/self-hosting
    - app/guides/streaming
    - app/api-reference/config/next-config-js/assetPrefix
---

Next.js sets standard `Cache-Control` headers that CDNs can use to cache responses at the edge. This page covers what works today, where CDN caching is challenging, and the direction toward eliminating custom-header dependencies.

## What Works Today

### Cache-Control headers

Next.js sets `Cache-Control` headers based on the rendering strategy of each route:

- **Static pages** (no revalidation): `s-maxage=31536000` (one year)
- **ISR pages** (time-based revalidation): `s-maxage={revalidate}, stale-while-revalidate={expire - revalidate}`. The default `expire` is one year, so `stale-while-revalidate` is included in the response header by default. You can customize this with [`cacheLife`](/docs/app/api-reference/functions/cacheLife).
- **Dynamic pages** (no caching): `private, no-cache, no-store, max-age=0, must-revalidate`

> CDNs that respect `s-maxage` and `stale-while-revalidate` can cache static and ISR pages at the edge. However, CDN-level caching alone does not support on-demand revalidation ([`revalidateTag()`](/docs/app/api-reference/functions/revalidateTag) / [`revalidatePath()`](/docs/app/api-reference/functions/revalidatePath)): those calls invalidate the Next.js server cache, but the CDN will continue serving its cached copy until the `s-maxage` TTL expires. To propagate on-demand revalidation to the CDN, wire up cache purging logic on the CDN side (for example, using the CDN's purge API when a revalidation event occurs).

Reviewer: This does not explain what wiring up the cache purging logic on the CDN side means, i.e. how revalidateTag would call a platform API or such or if it's meant that next to revalidateTag you also call the cdn api.

Reviewer 2: Updated with concrete purge flow: revalidate in Next.js, then purge affected HTML/RSC variants via CDN API.

### Static assets

> Static assets (JavaScript, CSS, images, fonts) served from `/_next/static/` include content hashes in their filenames and can be cached indefinitely with a long `max-age` and `immutable` directive.

Reviewer: ```suggestion
Static assets (JavaScript, CSS, images, fonts) served from `/_next/static/` include content hashes in their filenames and have a 1 year `max-age` and `immutable` directive: `public,max-age=31536000,immutable`
```

You can use [`assetPrefix`](/docs/app/api-reference/config/next-config-js/assetPrefix) to serve static assets from a different domain or CDN origin.

### Static prefetches (PPR-enabled routes)

When a route has Partial Prerendering enabled and the `next-router-prefetch` header is set (indicating a static prefetch), the response is deterministic: it returns the same prerendered content regardless of the client's router state. The `next-router-state-tree` header is not parsed for these requests, so it does not affect the response.

For PPR-enabled routes, a CDN can cache static prefetch responses if it:

1. Includes the `_rsc` search parameter in the cache key (to distinguish prefetch variants from HTML responses).
2. Respects the `Cache-Control` headers Next.js sets on the response.

> > **Good to know:** For routes without PPR, the `next-router-state-tree` header is still read during prefetch requests to determine which segments to include, which increases cache variability. The pathname-based direction described below removes this dependency for all routes. For segment-level prefetches, Next.js already uses pathname-based routing internally (e.g., `/page.segments/_tree.segment.rsc`), which CDNs can cache using standard pathname-based cache keys.

Reviewer: ```suggestion
> **Good to know:** For routes without PPR, the `next-router-state-tree` header is read during prefetch requests to determine which segments to include, which increases cache `vary` as it passes the current router state. For segment-level prefetches pathnames are used (e.g., `/page.segments/_tree.segment.rsc`). CDNs can cache these using standard pathname-based cache keys.
```

## Where CDN Caching Is Challenging

App Router responses can vary based on several custom request headers. Next.js sets a `Vary` header on responses to signal this to CDNs:

> - `rsc` — whether the request is a React Server Components (RSC) request

Reviewer: ```suggestion
- `rsc` — whether the request should return a React Server Components (RSC) payload instead of HTML
```

- `next-router-state-tree` — the client's current router state, used for targeted segment updates during dynamic navigations
- `next-router-prefetch` — whether this is a prefetch request
- `next-router-segment-prefetch` — the specific segment being prefetched
- `next-url` — added only for routes that use [interception routes](/docs/app/api-reference/file-conventions/intercepting-routes), carries the URL being intercepted

> In practice, many CDNs don't fully support `Vary` on custom headers. Next.js addresses this today with the `_rsc` search parameter: a hash of the relevant request header values that acts as a cache-key discriminator, ensuring different response variants get different cache keys. This ensures correct responses even on CDNs that ignore `Vary`, but it means CDNs that key on the full URL (including search params) see high cache variability.

Reviewer: ```suggestion
Many CDNs don't support `Vary` without additional configuration. Next.js addresses this with the `_rsc` search parameter: a hash of the relevant request header values that acts as a cache-key, ensuring different response variants get different cache keys. This ensures correct responses even on CDNs that ignore `Vary`.

## Handling Headers at the CDN

> ### What you can safely drop

Reviewer: ```suggestion
### What you can safely ignore
```

> These headers can be omitted without causing errors. The server returns a valid response, just potentially larger or less targeted:

Reviewer: ```suggestion
These headers can be omitted without causing errors. The server returns a valid response, just potentially larger or less targeted to the specific navigation:
```

> **`next-router-state-tree`**, **`next-router-prefetch`**, **`next-router-segment-prefetch`**: if omitted, the server returns a full RSC payload instead of a targeted segment update or lightweight prefetch. For static and ISR pages, the response is byte-identical regardless of these headers. For dynamic pages, the response is larger but correct.

Reviewer 3: You can never omit the cache busting param when `next-router-prefetch` is set. Treat `_rsc` as sacrosanct when that header is set. The client router will omit anything unnecessary from that hash already. So even though the hashes look highly entropic, they already represent the minimal set of inputs that contribute to the cachability of the response.

(Huge caveat: this isn't 100% true yet because we also put `Next-Url` in there, but we should just remove it from the cache busting param before we publish this advice.)

The only scenario where omitting the headers or dropping `_rsc` is allowed is during a dynamic request when `next-router-prefetch` is missing. But that's only relevant if you're choosing the cache dynamic responses, which is not idiomatic Next.js. However, if you want to do so anyway despite the caveats, then it's OK to omit `next-router-state-tree`; the server will still receive a valid response that client can understasnd.

Reviewer 2: Updated: explicitly states that for prefetch flows, next-router-prefetch and _rsc must both be preserved.

> **`next-router-state-tree`**, **`next-router-prefetch`**, **`next-router-segment-prefetch`**: if omitted, the server returns a full RSC payload instead of a targeted segment update or lightweight prefetch. For static and ISR pages, the response is byte-identical regardless of these headers. For dynamic pages, the response is larger but correct.

Reviewer: ```suggestion
**`next-router-state-tree`**, **`next-router-prefetch`**, **`next-router-segment-prefetch`**: if omitted, the server returns a full RSC payload instead of a targeted segment update or lightweight prefetch. For static and ISR pages, the response is identical regardless of these headers. For dynamic pages, the response is larger but still works for navigations.
```

> **`next-url`**: used for [interception routes](/docs/app/api-reference/file-conventions/intercepting-routes) to vary the response based on the referring page. If omitted, the response degrades to the non-intercepted state: the user sees the regular page instead of the intercepted modal. The response is still valid.

Reviewer 4: 🤔 It would be helpful to define "still valid" here. Is this saying the user-facing behaviour would be incorrect but the response would... just not be malformed?

Reviewer: Does [this update](https://github.com/vercel/next.js/pull/91237/changes#r2946881599) clarify the question? If next-url is omitted interception routes is not supported, as the server can't decide what the path you're navigating from is. So it can't resolve interception.

Reviewer 2: Updated wording to clarify "valid": server still returns a parseable response, with behavior degrading to less-targeted/non-intercepted results.

> **`next-url`**: used for [interception routes](/docs/app/api-reference/file-conventions/intercepting-routes) to vary the response based on the referring page. If omitted, the response degrades to the non-intercepted state: the user sees the regular page instead of the intercepted modal. The response is still valid.

Reviewer: ```suggestion
**`next-url`**: used for [interception routes](/docs/app/api-reference/file-conventions/intercepting-routes) to vary the response based on the referring page. If omitted, interception routes are not supported as the server doesn't know what original path to match against. The response returned is for regular navigation when `next-url` is omitted: the user sees the target page instead of the intercepted target page.
```

### What you must preserve

> **The `rsc` header** must be forwarded from the client to the server. This header tells the server to return an RSC payload instead of HTML. If a CDN strips it, the server returns HTML when the client-side router expects RSC data, which breaks client-side navigation. The `Vary` header and `_rsc` parameter exist specifically to prevent CDNs from serving a cached HTML response to an RSC request (or vice versa).

Reviewer: ```suggestion
**The `rsc` header** must be forwarded from the client to the server. This header tells the server to return an RSC payload instead of HTML. If a CDN strips it, the server returns HTML when the client-side router expects RSC data, which breaks client-side navigation, causing browser navigations instead. The `Vary` header and `_rsc` parameter exist specifically to prevent CDNs from serving a cached HTML response to an RSC request (or vice versa).
```

> **The `_rsc` search parameter** must be included in the cache key. It distinguishes response variants (HTML vs. RSC, different prefetch types). If an RSC request arrives without the correct `_rsc` value, the server responds with a **307 redirect** to the URL with the correct hash. CDNs should either follow this redirect or pre-populate `_rsc` to avoid double-request overhead. Ensure your CDN does not strip query parameters that look like cache-busting hashes, as some CDNs do this by default.

Reviewer: ```suggestion

**The `_rsc` search parameter** must be included in the cache key. It distinguishes response variants (HTML vs. RSC, different prefetch types). Ensure your CDN does not strip query parameters from cache keys, as some CDNs do this by default. When the `experimental.validateRSCRequestHeaders` option is enabled and a RSC request arrives without the correct `_rsc` value, the server responds with a **307 redirect** to the URL with the correct hash. CDNs should either follow this redirect or pre-populate `_rsc` to avoid double-request overhead.
```

> **The `_rsc` search parameter** must be included in the cache key. It distinguishes response variants (HTML vs. RSC, different prefetch types). If an RSC request arrives without the correct `_rsc` value, the server responds with a **307 redirect** to the URL with the correct hash. CDNs should either follow this redirect or pre-populate `_rsc` to avoid double-request overhead. Ensure your CDN does not strip query parameters that look like cache-busting hashes, as some CDNs do this by default.

Reviewer: Unclear to me what this means:

> or pre-populate `_rsc`

Reviewer 2: Updated: replaced unclear "pre-populate _rsc" wording with concrete behavior (follow redirect, or compute/rewrite _rsc upstream).

> **The `_rsc` search parameter** must be included in the cache key. It distinguishes response variants (HTML vs. RSC, different prefetch types). If an RSC request arrives without the correct `_rsc` value, the server responds with a **307 redirect** to the URL with the correct hash. CDNs should either follow this redirect or pre-populate `_rsc` to avoid double-request overhead. Ensure your CDN does not strip query parameters that look like cache-busting hashes, as some CDNs do this by default.

Reviewer 2: Updated: the `_rsc` mismatch 307 is now documented as conditional on experimental.validateRSCRequestHeaders, not default behavior.

> **Good to know:** Today, `next-url` is included in the `_rsc` hash even during static prefetches. This means you cannot safely ignore it under the current scheme without potentially getting cache misses. The pathname-based direction described below resolves this gap.

> ## Direction: Pathname-Based Cache Keying

Reviewer 4: Documenting future vision inline like this is a really interesting idea!

The Next.js team is working on moving all cache-affecting inputs into the URL pathname, eliminating the need for `Vary` on custom headers and removing the `_rsc` search parameter. This resolves the CDN caching challenges described above.

### How it works

The approach extends the routing scheme that [`output: 'export'`](/docs/app/guides/static-exports) and segment prefetches already use today. File extensions in the pathname identify the response type:

- **Full page RSC**: `/my/page.rsc` returns the RSC payload for the entire page
- **Segment RSC**: `/my/page.segments/path/to/segment.segment.rsc` returns the RSC payload for a specific segment

Under this model:

- **The pathname determines the cache key.** Anything in the pathname affects which response variant is returned.

> - **Search parameters can be safely dropped** without affecting response correctness.

Reviewer: ```suggestion
- **Search parameters can be safely dropped** without affecting returned responses.
```

- **Standard HTTP cache headers** (`Cache-Control`, `max-age`, etc.) are respected as usual.

> - **No custom header logic is required** from the CDN.

Reviewer: ```suggestion
- **No `Vary` support needed** from the CDN.
```

A CDN would cache Next.js responses by using the pathname as the cache key, ignoring search parameters, and respecting standard `Cache-Control` headers. No need to understand `Vary`, inspect custom headers, or program edge logic.

### What changes for interception routes

Under the current scheme, `next-url` contributes to the `_rsc` hash, so dropping it causes cache misses. Under the pathname-based scheme, interception variability would be encoded in a search parameter (not the pathname):

> - If a CDN preserves search params, interception would work correctly.

Reviewer: ```suggestion
- If a CDN preserves search params, interception works correctly.
```

> - If a CDN drops search params, interception would degrade to the non-intercepted state, but the response would still be valid.

Reviewer: ```suggestion
- If a CDN drops search params, interception is not supported. It would gracefully degrade to the non-intercepted page, client-side navigations won't break.
```

This makes interception route support an opt-in CDN capability rather than a requirement.

### Current status

> This direction extends patterns that are already operational in the codebase (segment prefetch paths, `output: 'export'` mode). It is in active design and does not have a ship date.

Reviewer: Since this is docs it has to be evergreen so calling out specific time in the future has high likelyhood of being missed when releasing.

```suggestion
This direction extends patterns that are already operational in the codebase (segment prefetch paths, `output: 'export'` mode). It is in active design.
```

> This direction extends patterns that are already operational in the codebase (segment prefetch paths, `output: 'export'` mode). It is in active design and does not have a ship date.

Reviewer: Specifically worth calling out where it's at. With segment paths already being used when cache components is enabled.

Reviewer 2: Updated: explicitly called out current segment-path usage when Cache Components is enabled.
